Revise SDL-0220: Support for Android custom RouterService (#695)
smartdevicelink/sdl_evolution: proposals/0220-support-for-android-custom-routerservice.md

    </service>
```

RouterService needs to communicate with every SDL application on device, so it's important to see how and when a SDL application starts proxy.

### How and when SDL apps start proxy
There are two major triggers where SDL applications start SdlProxy; one is a broadcast intent sent from RouterService, and the other is all other cases where the app starts SdlProxy by itself.
Let's see those two triggers in detail.

#### A trigger sent from RouterService

A SDL app usually starts proxy when the app receives START_ROUTER_SERVICE_ACTION, which is a broadcast intent sent from RouterService. In this case, the broadcast intent includes ComponentName of the RouterService, which is taken by SdlBroadcastReceiver, set it to an instance variable (called queuedService) and then onSdlEnabled is called.
The SDL app is responsible for overriding onSdlEnabled, starts its own service, and then starts Proxy. The Proxy checks to see queuedService, and uses that ComponentName to determine the target RouterService to bind with.

> This case works fine regardless of the RouterService is custom or not.

Reviewer: ```suggestion
This case works fine regardless of whether or not the RouterService is custom.
```

#### All other triggers on demand

> This refers to all other cases than "the trigger sent from RouterService". This case requires the app to determine which RouterService to bind with.

Reviewer: ```suggestion
This refers to all cases other than "the trigger sent from RouterService". This case requires the app to determine which RouterService to bind with.
```

The problem is that custom RouterService is unlikely chosen in step #2.
Step #1 actually depends on the timing of when the app calls queryForConnectedService. It is up to the application when queryForConnectedService is called, but [the integration-basic document](https://smartdevicelink.com/en/guides/android/getting-started/integration-basics/) suggests it should be called at Main Activities onCreate.

#### The cases where this proposal focuses on
The following section focuses on the case other than "a trigger sent from RouterService". For example, in the following cases, the app will start proxy:
- When OnLanguageChange is notified: in this case, the app has to restart the proxy.

> - When a user removes the app from Recent List: in this case, ActvityManager may automatically restarts the killed service, and restarts proxy in there. RouterService internally uses DeathRecipient interface to detect binderDied case, which might help some cases, but not all the cases.

Reviewer: ```suggestion
- When a user removes the app from Recent List: in this case, ActivityManager may automatically restart the killed service, and restart proxy in there. RouterService internally uses DeathRecipient interface to detect binderDied case, which might help some cases, but not all the cases.
```

> - When a user enforces the app to stop by settings | Apps | force stop: in this case ActivityManager may automatically restarts the killed service as well as the remove from Recent list case.

Reviewer: ```suggestion
- When a user enforces the app to stop by settings | Apps | force stop: in this case ActivityManager may automatically restart the killed service as well as the remove from Recent list case.
```

> - An app may pay attention to some error case (e.g. onProxyClosed, etc), and then restarts the proxy to get recovered.

Reviewer: ```suggestion
- An app may pay attention to some error case (e.g. onProxyClosed, etc), and then restart the proxy to get recovered.
```

> - An app may restart proxy when yet another transport is connected. This might be useful especially if the SDL core does not support secondary transport. For instance, if BT transport is connected, the app uses NON_MEDIA layout. When user connected USB or WIFI, then the app restarts the proxy instance and then uses NAV_FULLSCREEN_MAP layout.

Reviewer: ```suggestion
- An app may restart proxy when yet another transport is connected. This might be useful especially if the SDL core does not support secondary transport. For instance, if BT transport is connected, the app uses NON_MEDIA layout. When user connects USB or WIFI, then the app restarts the proxy instance and then uses NAV_FULLSCREEN_MAP layout.
```
